diff --git a/make/devkit/Tools.gmk b/make/devkit/Tools.gmk
index 5c29ee4db75..d4f4ce17721 100644
--- a/make/devkit/Tools.gmk
+++ b/make/devkit/Tools.gmk
@@ -62,13 +62,30 @@ ifeq ($(BASE_OS), OL)
     BASE_URL := https://yum.oracle.com/repo/OracleLinux/OL6/4/base/$(ARCH)/
     LINUX_VERSION := OL6.4
   endif
+else ifeq ($(BASE_OS), Centos)
+  DEFAULT_OS_VERSION := 7
+   ifeq ($(BASE_OS_VERSION), )
+     BASE_OS_VERSION := $(DEFAULT_OS_VERSION)
+   endif
+  BASE_OS_MAJOR_VERSION := $(shell echo $(BASE_OS_VERSION) | cut -d'.' -f1)
+  CENTOS_MIRROR := vault.centos.org
+  ifeq ($(ARCH), x86_64)
+    BASE_URL := http://$(CENTOS_MIRROR)/centos/$(BASE_OS_VERSION)/os/$(ARCH)/Packages/
+  else
+    BASE_URL := http://$(CENTOS_MIRROR)/altarch/$(BASE_OS_VERSION)/os/$(ARCH)/Packages/
+  endif
+  LINUX_VERSION := Centos$(BASE_OS_VERSION)
 else ifeq ($(BASE_OS), Fedora)
   ifeq ($(ARCH), riscv64)
     DEFAULT_OS_VERSION := rawhide/68692
     ifeq ($(BASE_OS_VERSION), )
       BASE_OS_VERSION := $(DEFAULT_OS_VERSION)
     endif
-    BASE_URL := http://fedora.riscv.rocks/repos-dist/$(BASE_OS_VERSION)/$(ARCH)/Packages/
+    ifeq ($(BASE_OS_VERSION), 28)
+      BASE_URL := https://secondary.fedoraproject.org/pub/alt/risc-v/archive/RPMS/riscv64
+    else
+      BASE_URL := http://fedora.riscv.rocks/repos-dist/f$(BASE_OS_VERSION)/latest/$(ARCH)/Packages/
+    endif
   else
     DEFAULT_OS_VERSION := 27
     LATEST_ARCHIVED_OS_VERSION := 35
@@ -188,6 +205,21 @@ endif
 
 ccache_ver_only := $(patsubst ccache-%,%,$(ccache_ver))
 
+# Centos GPG KEYS
+Centos_7_GPG_KEY_noarch := 6341AB2753D78A78A7C27BB124C6A8A7F4A80EB5  #gitleaks:allow
+Centos_7_GPG_KEY_aarch64 := EF8F3CA66EFDF32B36CDADF76C7CB6EF305D49D6 #gitleaks:allow
+Centos_7_GPG_KEY_x86_64 := 6341AB2753D78A78A7C27BB124C6A8A7F4A80EB5  #gitleaks:allow
+Centos_7_GPG_KEY_ppc64le := BAFA3436FC50768E3C3C2E4EA963BBDBF533F4FA #gitleaks:allow
+# Fedora GPG KEYS
+Fedora_19_GPG_KEY_s390x := CA81B2C85E4F4D4A1A3F723407477E65FB4B18E6   #gitleaks:allow
+
+# PGP Signature file extensions
+GCC_SIG := sig
+BINUTILS_SIG := sig
+CCACHE_SIG := asc
+GMP_SIG := sig
+MPC_SIG := sig
+GDB_SIG := sig
 
 GCC := http://ftp.gnu.org/pub/gnu/gcc/$(gcc_ver)/$(gcc_ver).tar.xz
 BINUTILS := http://ftp.gnu.org/pub/gnu/binutils/$(binutils_ver).tar.gz
@@ -255,7 +287,17 @@ download-rpms:
         # Only run this if rpm dir is empty.
         ifeq ($(wildcard $(DOWNLOAD_RPMS)/*.rpm), )
 	  cd $(DOWNLOAD_RPMS) && \
-	      wget -r -np -nd $(patsubst %, -A "*%*.rpm", $(RPM_LIST)) $(BASE_URL)
+	    wget -e robots=off -r -np -nd $(patsubst %, -A "*%*.rpm", $(RPM_LIST)) $(BASE_URL)
+	    gpg --keyserver keyserver.ubuntu.com --recv-keys \
+	      $($(BASE_OS)_$(BASE_OS_MAJOR_VERSION)_GPG_KEY_$(ARCH))
+	    gpg --keyserver keyserver.ubuntu.com --recv-keys \
+	      $($(BASE_OS)_$(BASE_OS_MAJOR_VERSION)_GPG_KEY_noarch)
+	    rm -f $(DOWNLOAD_RPMS)/rpm.sig && gpg --armor --export \
+	      $($(BASE_OS)_$(BASE_OS_MAJOR_VERSION)_GPG_KEY_$(ARCH)) > $(DOWNLOAD_RPMS)/rpm.sig
+	    gpg --armor --export \
+	      $($(BASE_OS)_$(BASE_OS_MAJOR_VERSION)_GPG_KEY_noarch) >> $(DOWNLOAD_RPMS)/rpm.sig
+	    rpm --import $(DOWNLOAD_RPMS)/rpm.sig || true
+	    rpm -K $(DOWNLOAD_RPMS)/*.rpm
         endif
 
 ##########################################################################################
@@ -280,6 +322,14 @@ define Download
 
   $$($(1)_FILE) :
 	wget -P $(DOWNLOAD) $$($(1))
+        ifeq ($(1),MPFR)
+	  echo $($(1)_SHA256) $$@ | sha256sum -c
+        else
+	  wget -P $(DOWNLOAD) $$($(1)).$($(1)_SIG)
+	  gpg --keyserver keyserver.ubuntu.com --recv-keys $($(1)_GPG_KEY)
+	  echo -e "5\ny\n" | gpg --batch --command-fd 0 --expert --edit-key $($(1)_GPG_KEY) trust;
+	  gpg --verify $$@.$($(1)_SIG) $$@
+        endif
 endef
 
 # Download and unpack all source packages
@@ -332,6 +382,9 @@ $(foreach p,$(RPM_FILE_LIST),$(eval $(call unrpm,$(p))))
 # have it anyway, but just to make sure...
 # Patch libc.so and libpthread.so to force linking against libraries in sysroot
 # and not the ones installed on the build machine.
+# Remove comment sections from static libraries and C runtime objects
+# to prevent leaking RHEL-specific package versions into
+# devkit-produced binaries.
 $(libs) : $(rpms)
 	@echo Patching libc and pthreads
 	@(for f in `find $(SYSROOT) -name libc.so -o -name libpthread.so`; do \
@@ -341,6 +394,7 @@ $(libs) : $(rpms)
 	      -e 's|/lib/||g' ) > $$f.tmp ; \
 	  mv $$f.tmp $$f ; \
 	done)
+	@find $(SYSROOT) -name '*.[ao]' -exec objcopy --remove-section .comment '{}' ';'
 	@mkdir -p $(SYSROOT)/usr/lib
 	@touch $@
 
@@ -449,6 +503,9 @@ endif
 
 # Makefile creation. Simply run configure in build dir.
 # Setting CFLAGS to -O2 generates a much faster ld.
+# Use --enable-deterministic-archives so that make targets that
+# generate "ar" archives, such as "static-libs-image", produce
+# deterministic .a files.
 $(bfdmakes) \
 $(BUILDDIR)/$(binutils_ver)/Makefile \
     : $(BINUTILS_CFG)
@@ -463,6 +520,7 @@ $(BUILDDIR)/$(binutils_ver)/Makefile \
 	      --with-sysroot=$(SYSROOT) \
 	      --disable-nls \
 	      --program-prefix=$(TARGET)- \
+	      --enable-deterministic-archives \
 	      --enable-multilib \
 	      --enable-threads \
 	      --enable-plugins \
@@ -557,6 +615,7 @@ $(BUILDDIR)/$(gcc_ver)/Makefile \
 	      --with-mpfr=$(PREFIX) \
 	      --with-gmp=$(PREFIX) \
 	      --with-mpc=$(PREFIX) \
+	      --enable-linker-build-id \
 	) > $(@D)/log.config 2>&1
 	@echo 'done'
 
